clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp

```diff
@@ -264,44 +244,82 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
+  auto BitcodeOutput =
+      createTempFile(Args, sys::path::filename(OutputFile), "bc");
+  if (!BitcodeOutput)
+    return BitcodeOutput.takeError();
+
+  // Get all SYCL device library files, if any
   auto SYCLDeviceLibFiles = getSYCLDeviceLibs(Args);
   if (!SYCLDeviceLibFiles)
     return SYCLDeviceLibFiles.takeError();
-  if ((*SYCLDeviceLibFiles).empty())
-    return InputFile;
 
-  Expected<std::string> LLVMLinkPath =
-      findProgram(Args, "llvm-link", {getMainExecutable("llvm-link")});
-  if (!LLVMLinkPath)
-    return LLVMLinkPath.takeError();
+  if (Verbose) {
+    std::string Inputs =
+        std::accumulate(std::next(InputFiles.begin()), InputFiles.end(),
+                        InputFiles.front(), [](std::string a, std::string b) {
+                          return std::move(a) + ", " + std::move(b);
+                        });
+    std::string LibInputs = "";
+    if (!(*SYCLDeviceLibFiles).empty())
+      LibInputs = std::accumulate(
+          std::next((*SYCLDeviceLibFiles).begin()), (*SYCLDeviceLibFiles).end(),
+          (*SYCLDeviceLibFiles).front(), [](std::string a, std::string b) {
+            return std::move(a) + ", " + std::move(b);
+          });
+    errs() << formatv(
+        "sycl-device-link: inputs: {0} libfiles: {1} output: {2}\n", Inputs,
+        LibInputs, *BitcodeOutput);
+  }
 
-  // Create a new file to write the linked device file to.
-  auto OutFileOrErr =
-      createTempFile(Args, sys::path::filename(OutputFile), "bc");
-  if (!OutFileOrErr)
-    return OutFileOrErr.takeError();
+  LLVMContext C;
+  auto LinkerOutput = std::make_unique<Module>("sycl-device-link", C);
+  Linker L(*LinkerOutput);
+  // Link SYCL device input files.
+  for (auto &File : InputFiles) {
+    auto ModOrErr = getBitcodeModule(File, C);
+    if (!ModOrErr)
+      return ModOrErr.takeError();
+    if (L.linkInModule(std::move(*ModOrErr)))
+      return createStringError("Could not link IR");
+  }
 
-  SmallVector<StringRef, 8> CmdArgs;
-  CmdArgs.push_back(*LLVMLinkPath);
-  CmdArgs.push_back("-only-needed");
-  CmdArgs.push_back(InputFile);
-  for (auto &File : *SYCLDeviceLibFiles)
-    CmdArgs.push_back(File);
-  CmdArgs.push_back("-o");
-  CmdArgs.push_back(*OutFileOrErr);
-  CmdArgs.push_back("--suppress-warnings");
-  if (Error Err = executeCommands(*LLVMLinkPath, CmdArgs))
-    return std::move(Err);
-  return *OutFileOrErr;
+  // Link in SYCL device library files.
+  const llvm::Triple Triple(Args.getLastArgValue(OPT_triple));
+  for (auto &File : *SYCLDeviceLibFiles) {
+    auto LibMod = getBitcodeModule(File, C);
+    if (!LibMod)
+      return LibMod.takeError();
+    if ((*LibMod)->getTargetTriple() == Triple) {
+      unsigned Flags = Linker::Flags::LinkOnlyNeeded;
+      if (L.linkInModule(std::move(*LibMod), Flags))
+        return createStringError("Could not link IR");
+    }
+  }
+
+  if (Args.hasArg(OPT_print_linked_module))
+    errs() << *LinkerOutput;
```

Comment:
I don't understand the reason to print module to the error stream. LIT will fail is tests writes to the error stream. Redirecting the error stream to the output stream each for each test is annoying. Usually, you want your error stream to be empty. Non-empty error stream should signal about errors. I think other tools like clang driver print out verbose non-error output to the output stream.
We should discuss `clang-linker-wrapper` choice in the upstream and if `clang-sycl-linker` should follow the same approach.

```suggestion
    outs() << *LinkerOutput;
```

Reply:
Interesting. Will change it to outs(). Thanks